File: src/ecatvasp/desktop/site_preflight.py
"""Desktop-facing action for transient real-site preflight diagnostics."""

from __future__ import annotations

from typing import Any

from ecatvasp.api.site_preflight import SitePreflightApplicationService
from ecatvasp.domain import SchedulerType
from ecatvasp.execution.preflight import PreflightCheck, PreflightReport
from ecatvasp.execution.site_profile import SiteProfile, SshMode


def site_preflight_action(
    profile: dict[str, Any],
    *,
    service: SitePreflightApplicationService | None = None,
) -> dict[str, object]:
    """Run preflight from a user-local profile and return sanitized transient evidence."""

    selected_service = SitePreflightApplicationService() if service is None else service
    report = selected_service.run(_site_profile(profile))
    return _report_payload(report)


def _site_profile(raw: dict[str, Any]) -> SiteProfile:
    modules_raw = raw.get("module_loads", [])
    if not isinstance(modules_raw, list):
        raise ValueError("profile.module_loads must be a string list")
    return SiteProfile(
        site_id=str(raw["site_id"]),
        name=str(raw["name"]),
        host_alias=str(raw["host_alias"]),
        remote_root=str(raw["remote_root"]),
        potcar_resolver_id=str(raw["potcar_resolver_id"]),
        potcar_family=str(raw["potcar_family"]),
        potcar_root=str(raw["potcar_root"]),
        scheduler_type=SchedulerType(str(raw.get("scheduler_type", SchedulerType.SLURM.value))),
        ssh_mode=SshMode(str(raw.get("ssh_mode", SshMode.SYSTEM_OPENSSH.value))),
        vasp_executable=str(raw.get("vasp_executable", "vasp_std")),
        mpi_launcher=_optional_text(raw.get("mpi_launcher")),
        module_loads=tuple(str(item) for item in modules_raw),
        bader_executable=_optional_text(raw.get("bader_executable")),
        lobster_executable=_optional_text(raw.get("lobster_executable")),
    )


def _report_payload(report: PreflightReport) -> dict[str, object]:
    return {
        "site_id": report.site_id,
        "profile_hash": report.profile_hash,
        "status": report.status.value,
        "timestamp": report.timestamp,
        "transient": True,
        "checks": [_check_payload(check) for check in report.checks],
    }


def _check_payload(check: PreflightCheck) -> dict[str, object]:
    return {
        "check_name": check.check_name,
        "status": check.status.value,
        "reason_code": None if check.reason_code is None else check.reason_code.value,
        "message": check.message,
        "evidence": list(check.evidence),
    }


def _optional_text(value: object) -> str | None:
    if value is None:
        return None
    if not isinstance(value, str):
        raise ValueError("optional site-profile command fields must be strings")
    return value


__all__ = ["site_preflight_action"]

File: ui/desktop/src/lib/backend/site-preflight.ts
import { DesktopContractError } from "./contracts";

export type SitePreflightStatus = "READY" | "BLOCKED" | "WARNING" | "UNAVAILABLE_OPTIONAL";

export interface SiteProfileInput {
  site_id: string;
  name: string;
  host_alias: string;
  remote_root: string;
  potcar_resolver_id: string;
  potcar_family: string;
  potcar_root: string;
  scheduler_type?: "slurm";
  ssh_mode?: "system_openssh";
  vasp_executable?: string;
  mpi_launcher?: string;
  module_loads?: string[];
  bader_executable?: string;
  lobster_executable?: string;
}

export interface SitePreflightCheck {
  check_name: string;
  status: SitePreflightStatus;
  reason_code: string | null;
  message: string;
  evidence: string[];
}

export interface SitePreflightPayload {
  site_id: string;
  profile_hash: string;
  status: SitePreflightStatus;
  timestamp: string;
  transient: true;
  checks: SitePreflightCheck[];
}

const STATUSES = new Set<SitePreflightStatus>([
  "READY",
  "BLOCKED",
  "WARNING",
  "UNAVAILABLE_OPTIONAL",
]);
const SHA256 = /^[0-9a-f]{64}$/;

export function assertSitePreflightPayload(
  payload: SitePreflightPayload,
  expectedSiteId: string,
): void {
  const value: unknown = payload;
  if (!isRecord(value) || value.site_id !== expectedSiteId) {
    throw new DesktopContractError("site preflight identity is invalid");
  }
  if (typeof value.profile_hash !== "string" || !SHA256.test(value.profile_hash)) {
    throw new DesktopContractError("site preflight profile hash is invalid");
  }
  if (!isStatus(value.status)) {
    throw new DesktopContractError("site preflight status is invalid");
  }
  if (typeof value.timestamp !== "string" || value.timestamp.length === 0 || value.transient !== true) {
    throw new DesktopContractError("site preflight transient evidence metadata is invalid");
  }
  if (!Array.isArray(value.checks) || !value.checks.every(isCheck)) {
    throw new DesktopContractError("site preflight checks are invalid");
  }
}

function isCheck(value: unknown): value is SitePreflightCheck {
  if (!isRecord(value)) return false;
  return (
    typeof value.check_name === "string" &&
    value.check_name.length > 0 &&
    isStatus(value.status) &&
    (value.reason_code === null || typeof value.reason_code === "string") &&
    typeof value.message === "string" &&
    Array.isArray(value.evidence) &&
    value.evidence.every((item) => typeof item === "string")
  );
}

function isStatus(value: unknown): value is SitePreflightStatus {
  return typeof value === "string" && STATUSES.has(value as SitePreflightStatus);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
